[Bongf] 미션4 : 모든 기물 배치하기

bong6981 opened the pull request:
안녕하세요! 

이번에 고민했던 & 변경했던 부분 말씁드립니다.
1. enum Color와 Representation 을 어디에 둘 것인가 고민했습니다.
- 기존에는 Pawn 밖에 Color를 두었습니다.
- Pawn이 Piece로 변경되면서 Piece 안에 두었습니다.  
- Piece 안에 두면 너무 지저분해지는 것 같았습니다. 
- Piece 밖으로 뺄까 고민했지만 Piece 외에 쓰일 곳이 없을 것 같았습니다. 

2. 여전히 해결되지 않는 접근제어자 
- Board클래스에 addWhite/Black Pawn/Piece() 메서드의 경우 private으로 선언하고 싶었지만, 
- 해당 인자들의 조건을 체크해주고 예외처리를 해주는 부분이 있어 test를 해줘야겠다고 생각했습니다. 
- test하다보니 private -> default 로 접근제어자를 변경하게 되었고 test 코드 때문에 접근제어자를 바꾸는 지난 번 미션 때의 고민에 다시 봉착하게 되었습니다. 
-  왜 private한 접근제어자가 왜 test되어야 하는지 생각해 보라고 하셨는데, 제가 개념 이해를 잘못한 것인지 해결되지 않았습니다. 

 코드리뷰 감사하다는 말씀 드립니다.

bong6981 committed:
feat : Add StringUtilsTest, StringUtils

- StringUtils는 NEW_LInew(개행 문자열)을 상수로 갖고 있습니다
- StringUtils의 appendNewLine 메소드는 인자로 전달받는 String에 개행을
  추가합니다.
- StringUtils를 활용하는 방식으로 Board의 print()메소드를 수정했습니다.

bong6981 committed:
feat : Add methods to create pieces of each type

- Pawn 클래스 이름을 Piece로 rename
- Piece로 모든 기물을 포함하는 개념으로 변함에 따라, Color가 Piece에
  종속적이라 생각하여를 내부 enum 클래스로 변경
- representaion도 기물 종류별로 다른 값을 가지므로 Color에서 분리시킴
- Piece가 private 한 생성자를 가지도록 변경하고, 인스턴스 생성한
  이후에는 인스턴스 상태 값을 변경할 수 없도록 final로 선언
- 색과 종류별로 Piece객체를 생성할 수 있도록 팩토리메서드를 추가

bong6981 committed:
feat : Add BoardTest and fis Board

- 보드테스트를 전체 체스판의 상태를 테스트 할 수 있도록 작성
- 보드 클래스 테스트를 통과하도록 수정
- 현재 단계는 단순 기물 배치까지만이라 기물을 배치하는 4줄을 각각
  whitePieces, whitePawns, blackPawns, blackPieces로 관리함

bong6981 committed:
feat : Add isWhite() and isBlack()

- 색을 구분할 수 있도록 isWhite()와 isBlack()을 추가

bong6981 committed:
refacotr : Edit verifying code using isWhite/Black()

- 색 검증을 하는 부분을 isWhite/Black()을 이용하는 것으로 수정

bong6981 committed:
refactor : Edit code

- 자잘자잘한 부분에 수정에 있어 한 번에 리팩토링
- assertAll()로 처리가 주요 수정 내용

bong6981 committed:
refacotr : Rename method

- 리팩토링 과정에서(Pawn -> Piece) 반영되지 못한 메소드 부분을 piece로
  수정함

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -1,92 +1,126 @@
+        for (char representation : getRepresentations(pieces)) {
+            sb.append(representation);
         }
-        whitePawns.add(pawn);
+        return sb.toString();
     }
 
-    public void addBlackPawn(Pawn pawn) {
-        if (pawn.getColor() != Color.BLACK) {
-            throw new InvalidParameterException("pawn의 색을 확인해 주세요");
+    private char[] getRepresentations(List<Piece> pieces) {
+        char[] representations = new char[pieces.size()];
+        for (int i = 0; i < pieces.size(); i++) {
+            representations[i] = pieces.get(i).getRepresentation();
         }
-        blackPawns.add(pawn);
+        return representations;
     }
```
두 개 포문을 다음에는 스트림으로 바꿔서 해보시겠어요?ㅎㅎ

bong6981 replied:
다음 번에는 스트림을 활용하도록 해보겠습니다 !

bong6981 replied:
네 다음 번에는 스트림으로 바꿔 시도해보겠습니다!

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -1,92 +1,126 @@
 
-    public int size() {
-        return whitePawns.size() + blackPawns.size();
+    private List<Piece> whitePieces = new ArrayList<>();
+    private List<Piece> whitePawns = new ArrayList<>();
+    private List<Piece> blackPawns = new ArrayList<>();
+    private List<Piece> blackPieces = new ArrayList<>();
```
변수가 조금 많다고 생각이들지만 현재 요구사항에는 만족하기때문에 이대로 가 보시고,
조금 더 요구사항이 다양해지면 중복으로 보이는 이 부분을 어떻게 추상화 할 것인지 고민하셔야 할것같네요!

bong6981 replied:
네! 다음 미션에서는 리스트가 늘어나서 고민하고 있습니다.  추상화를 고민해보겠습니다.

bong6981 replied:
다음단계에서 만들어야 할 리스트가 더 많은 것을 확인했습니다. 추상화를 고민해 보겠습니다!

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -1,92 +1,126 @@
+
+    void addWhitePawn(Piece piece) {
+        if (!piece.isWhite() || piece.getType() != Piece.Type.PAWN) {
+            throw new InvalidParameterException("흰색의 pawn만 추가할 수 있습니다.");
```
InvalidParameterException 클래스는 java.security. 모듈에서 암호화 관련 로직을 짜다가 그들이 사용하려고 만든 Exception입니다. 이름만 그럴싸할 뿐 일반 개발자들이 가져다 쓰기는 좀 옷이 안맞죠..

java.lang에 IllegalArgumentException 등이 많이 쓰이고, 아니면 커스텀 Exception을 만들어 사용하시는것도 좋습니다~

bong6981 replied:
저번에도 받은 피드백이어서 수정했는데.. 왜 이렇게 되었는지 죄송합니다. 다시 수정하도록 하겠습니다.

LarryJung commented on src/main/java/pieces/Piece.java:
```diff
@@ -0,0 +1,108 @@
+        return (this.isWhite()) ? this.type.getWhiteRepresentation() : this.type.getBlackRepresentation();
+    }
+
+    public static Piece createWhitePawn() {
+        return new Piece(Color.WHITE, Type.PAWN);
+    }
+
```
혹시.. 싱글톤에 대해서 알고 계신다면 whitePawn, blackPawn 등을 싱글톤으로 만들면 어떤 장점/단점이 있을지 생각해보시는 것도 도움이 되실것 같습니다.

참고로 제 의견은 싱글톤으로 두지 않는게 더 낫다 입니다.ㅎㅎ

bong6981 replied:
싱글톤에 알고 있지 않아 짧게 학습해보니 간단하게 오지 하나의 인스턴스만 생성하게 해서 관리하게 하는 방법이라는 것을 알게 되었습니다 ( 해당 부분 더 공부해야 할 것 같습니다 ㅠㅠ ) 
- 아직 이해가 미흡하여 잘 이해가 가지 않습니다. 
- whitePawn/blackPawn을 싱글톤으로 만든다면.. 오직 1개의 객체만 생성할 수 밖에 없는 것 아닌가요...? 그렇다면 pawn은 각각 8개씩이어야 하는데 이는 불가하지 않을까 생각됩니다. (학습이 부족하여 말씀주신 질문을 잘 이해하지 못했습니다) 
- 만약 말씀하신 부분이 createWhitePiece의 각각의 기물들을 싱글톤으로 만든 것을 의도하신 거라면 
  -  장점은 정해진 말들만 생겨서 다른 객체가 생길 가능성을 막아준다. 
  -  단점은 멀티쓰레드에서 동기화 처리를 안하면 객체가 두 개 생길 수 있다는 점은 찾아 보았으나 지금 단게에서는 와닿지 않았습니다
    - 단점을 생각을 해보았을 때는 체스게임을 하면서 기물에 상태값의 변경이 필요한 경우 ( 나중에 기물에 위치를 속성으로 가진다든지) 기물이 변하는 상태를 갖게 되고 -> 전역에서 이에 접근할 수 있다면 데이터가 여기저기서 잘못 바꿔질 수 있지 않을까 하는 생각이듭니다... 
    -  또 생각해보았을 때, 체스게임을 하다보면 기물이 죽는 경우가 생기는데 이를 보드에서 단순히 내려오는 정도로만 구현하지 않을까 (싱글톤 객체 삭제 과정이 추가적으로 필요하지 않을까 생각했습니다) 그렇다면 메모리의 낭비가 아닐까 생각했습니다.     
- 제가 싱글톤에 대해서 아직 몰라 터무니없는 말들만 남긴다는 생각이듭니다. 더 학습해보겠습니다!

LarryJung replied:
사실 이 미션에 싱글톤의 개념이 꼭 필요한것은 아니지만.. 그냥 주저리주저리 해보겠습니다.ㅋㅋ

코드만 딱 봤을때 든 생각은 (비숍이든, 폰이든..)
```java
public static Piece createWhiteBishop() {
        return new Piece(Color.WHITE, Type.BISHOP);
    }
```
흰색비숍은 한번 객체 생성이 되면 더 이상 변할 여지도 없어 보이고, 똑같은 것을 찍어낼 바에 싱글톤으로 만드는게 메모리 비용을 아낄 수 있지 않을까? 생각이 들었거든요ㅎㅎ

예를들어 유저의 계좌잔액이 없다는 것을 `return Money.empty()` 이런식으로 응답하고 있고, 
100만명 중 90만명이 계좌 잔액이 없는데 잔액 조회 횟수가 아주아주 많다고 하면 매번 Money.empty() 로 객체를 매번 새로 생성하는게 아니라 `private Money emptyMoney = Money.empty();` 로 미리 만들어 놓고 `return emptyMoney` 만 해줘도 되겠죠?

근데 체스에서는 개념적으로 흰색 비숍이 두개가 필요하고, 두개의 메모리 주소를 다르게 가져가야 실수를 방지할 수 있을 것 같다는 생각때문에 싱글톤보다는 객체를 각각 생성해주는게 낫겠다라는 생각이 들었습니다.~ ㅎㅎ 

그냥 그렇다고만 생각해주세요 ㅋㅋ

bong6981 replied:
오! 네 조금 더 공부해보겠습니다~!

LarryJung commented on src/main/java/pieces/Piece.java:
```diff
@@ -0,0 +1,108 @@
+        return new Piece(Color.BLACK, Type.KING);
+    }
+
+    public boolean isBlack() {
+        return this.color == Color.BLACK;
+    }
```
enum 과 값 비교를 하는 일은 정말정말 많은데요, 팁을 드리면 enum이 먼저 오는게 더 좋습니다. 왜냐면 지금 경우는 this.color가 final 이고 notnull 이라서 상관이 없지만 만약 this.color 가 null 인 경우에는 NPE가 뜹니다. 반대로하면 false 가 나오구요. 

```java
return this.color == Color.BLACK; // throw NPE
```

```java
return Color.BLACK == this.color (null) // false
```

bong6981 replied:
오 해당 부분에 대해서는 전혀 모르고 있었습니다. 감사합니다!

LarryJung reviewed (commented):
안녕하세요~ 지난 미션에도 그렇고 큰 수정사항이 눈에 띄지 않게 잘 구현해주셨습니다. 

private 메서드와 테스트 필요성에 대해 의견만 장황하게 남겨봅니다.ㅎㅎ
저도 비슷한 고민을 했었던 것 같네요~

이번 요구사항을 TDD로 한다고 가정하면, 테스트 코드에 제일 먼저 작성할 기능이 showBoard() 겠죠?
그리고 세부 코드를 짜면서 리팩토링된 산물이 private 메서드로 나오게 될 것 같구요

그래서 테스트 코드는 요구사항을 직접적으로 만족할 기능에 먼저 집중합니다. 그게 public 메서드로 만들어졌죠..

지금 코드는 이전 미션에서 발전(?) 됐고, 그땐 폰 하나 추가하는 기능이 직접적인 요구사항이었기 때문에 당연히 addPawn 메서드는 public 이었을거에요.

하지만 지금은 Board를 가져다 쓰는 쪽(Main)에서 addPawn을 사용할 일도 없기 때문에 접근제어자를 최소화 하는게 맞아 보이죠, 
그런 상황에서 기능을 테스트하고 싶을때는 클래스를 추가하면 됩니다. 그리고 그 클래스의 접근제어자를 Board까지만 노출시키면 Main에서 임포트하지 못하기 때문에 안전하죠. 

```java
class BoardHelper {
  public static void addWhitePawn(Board board, Piece piece) {
    if (!piece.isWhite() ... ) ...
  }
}
```

예시 코드처럼 static으로 할 수도 있고, 경우에 따라서 객체를 생성해서 사용할 수 도 있습니다.

요지는 public메서드 만으로 private메서드의 테스트 케이스를 전부 커버하기 어려울 정도로 private메서드가 복잡해졌을때는 그 기능을 위임할 새로운 객체가 등장할 수 있는 사인이라고 생각해보셔도 될것 같네요~

LarryJung reviewed (changes requested):
request changes 마크용입니다. 제가 자꾸 빼먹어서^^ 😀

bong6981 committed:
fix: Edit code

- 피드백 반영하여 IllegalArgumentException으로 수정
- isWhite/Black()의 return문에서 enum이 먼저 올 수 있게 수정

bong6981 commented:
우선 말씀 주신 부분 수정했습니다. 
- IllegalArgumentException으로 수정하고, 
-  isWhite/Black() 의 return에서 this.color가 null인 경우를 대비해서 enum을 먼저 써주었습니다. 

---

그 외 말씀주신 싱글톤의 개념은 좀 더 학습해보겠습니다! 

접근 제어자에 대해서는 답변 주셔서 감사합니다! 
해당 고민에 봉착한 경우 이것이 public 메소드만으로 검사할 수 없는 test인가 
그렇다면 이를 별도의 클래스로 분리하는 것이 맞지 않는가 
이렇게 고민해보겠습니다.
더 학습해봐야겠지만 새로운 해결책이 열린 느낌입니다 👍  감사합니다

LarryJung reviewed (approved):
네 수고하셨습니다~ 머지하고 다음미션 진행하시죠!
